### PR TITLE
Switch to PackageCompiler.create_library()

### LINKED ISSUE #5: Support multi-versioning
Would likely hit https://github.com/JuliaLang/julia/issues/34064, could use workaround in https://github.com/JuliaLang/PackageCompiler.jl/blob/1e611e47500e74915cd854f621929db03da34871/src/PackageCompiler.jl#L625-L627.

### PR DESCRIPTION
**Pull request by kmsquire**
Fixes #5, #11

Depends on https://github.com/JuliaLang/PackageCompiler.jl/issues/490

This does make things slightly more complicated ***Edit: on Windows*** (it's harder to find the library at run time), ~so if you want to keep this repo  simple, then it may be better not to merge~.  

Either way, would love feedback.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -15,21 +23,32 @@ else
...
+ifeq ($(OS), Darwin)
+	# Make sure we can find and use the shared library on OSX
+	install_name_tool -change $(LIBCG) $(LIB_LIBCG) $@
+	## Alternatively, if we install the lib directory somewhere in the library
+	## search path, we could run
+	# install_name_tool -change $(LIBCG) @rpath/$(LIBCG) $@
+else ifeq ($(OS), WINNT)
+	echo "Please add $(LIBDIR) to your PATH before running $(MAIN)"
+else
+	echo "Please add $(LIBDIR) to LD_LIBRARY_PATH before running $(MAIN)"
```
Comment (jedbrown) [inline Makefile:50]: Can you link with `-Wl,-rpath,$(LIBDIR)` instead?

File: build/Manifest.toml [hunk 2]
```diff
@@ -25,22 +47,32 @@ uuid = "56ddb016-857b-54e1-b83d-db4d58db5568"
...
+[[MozillaCACerts_jll]]
+uuid = "14a3606d-f60d-562e-9121-12d972cd8159"
+
+[[NetworkOptions]]
+uuid = "ca575930-c2e3-43a9-ace4-1e988b2c1908"
+
 [[PackageCompiler]]
 deps = ["Libdl", "Pkg", "UUIDs"]
-git-tree-sha1 = "98aa9c653e1dc3473bb5050caf8501293db9eee1"
+path = "/Users/kevinsquire/.julia/dev/PackageCompiler"
```
Comment (simonbyrne) [inline build/Manifest.toml:62]: Need to `Pkg.add` by URL so that this will work

**Comment by simonbyrne** [comment]
Thanks @kmsquire! This looks very handy.

File: Makefile (hunk 1 above)
Comment (simonbyrne) [reply Makefile:50]: I agree, rpath seems like a cleaner way to do this.

Comment (kmsquire) [reply Makefile:50]: Thanks, wasn't really sure how to do this.  Will fix.

Comment (kmsquire) [reply Makefile:50]: (Although now that I look at it, it makes perfect sense...)

File: build/Manifest.toml (hunk 2 above)
Comment (kmsquire) [reply build/Manifest.toml]: Done.

File: Makefile (hunk 1 above)
Comment (kmsquire) [reply Makefile]: This is fixed for Linux (and OSX).  There doesn't seem to be an equivalent for Windows (according to the PackageCompiler documentation).

Comment (simonbyrne) [reply Makefile]: I think Windows will look for dlls in the same directory as the executable, so setting `LIBDIR := $(OUTDIR)` might be an easy fix

Comment (kmsquire) [reply Makefile]: Okay, that makes sense.

Comment (kmsquire) [reply Makefile]: I did add `$(LIBDIR)` to the RPATH, and put all of the library outputs next to the executable (in `target/bin`) on Windows.

**Comment by kmsquire** [comment]
Status update: this is working well (compiles and tests pass) on OSX and Linux.  However, I haven't been able to get the Windows binary to run. (See [here](https://github.com/simonbyrne/libcg/pull/18/checks?check_run_id=1787992553)--this only has Julia 1.6, but the same thing happens in 1.5.)

As can be seen in the link above, PackageCompiler is able to produce the library, and the main executable is compiled and linked fine.  Running `dir` on the bin directory shows `main.exe` and all of the `dll`s.  But when it's run, it always gives

```
ERROR: could not load library "D:\a\libcg\libcg\target"
The specified module could not be found. 
```

Nothing in the debug logs during make jumps out at me.  I don't have a Windows machine to test on locally, so I've been dependent on CI, and I'm out of ideas.  Does anyone have any ideas?  

(I turned off building for OSX and Linux CI while testing Windows.)

**Comment by simonbyrne** [comment]
@staticfloat: any ideas?

**Comment by staticfloat** [comment]
Is it trying to load a library called `target`?  No `.dll`, no nothing?

**Comment by kmsquire** [comment]
> Is it trying to load a library called target? No .dll, no nothing?

That's what the message seems to indicate.

The library (`cg.dll` is created [here](https://github.com/simonbyrne/libcg/pull/18/checks?check_run_id=1787992553#step:8:123), and `main.exe` is linked to the library [right below that](https://github.com/simonbyrne/libcg/pull/18/checks?check_run_id=1787992553#step:8:123), with

```
gcc -o D:/a/libcg/libcg/target/bin/main.exe main.o -lm -LD:/a/libcg/libcg/target/bin -ljulia  -lcg
```

I believe I've tried changing all of the forward slashes to backslashes, although I'm pretty sure it shouldn't matter (`main.exe` gets created fine, at least).

AFAICT, there isn't anything equivalent to `ldd` or `otool` for Windows installed.  I might try to get DependencyWalker installed and look at that.  

Is there an easy way to get up and running with Julia on Azure or some other remote Windows service where I could work through things manually?

(TBH, I've completed the work needed to get things working nicely with `PackageCompiler` under OSX and Linux, which are all I need for my own use cases.  But it would be nice to be complete, and I believe `PackageCompiler` itself works under Windows, so I don't see why this shouldn't.)

**Comment by staticfloat** [comment]
> AFAICT, there isn't anything equivalent to ldd or otool for Windows installed. 

It would be interesting to see what `ObjectFile.jl` says about it.  You can print out the dependencies via something like:

```
using ObjectFile
readmeta(path_to_dll) do oh
	@show path.(DynamicLinks(oh))
end
```

**Comment by kmsquire** [comment]
@simonbyrne thanks for the link to `reverse-rdp-windows-github-actions`--that looks like it should do the trick.  I'd like to try to use it, but it would be useful to get collaborator access to this repo so I can cancel CI jobs.  Would you mind giving me access?

**Comment by simonbyrne** [comment]
Done!

**Comment by kmsquire** [comment]
I also need to set up a secret (NGROK_AUTH_TOKEN).  Can I send that to you (maybe via Slack?) and ask you to set that for me under Settings -> Secrets -> Repository Secrets?

(Not sure if there's a more secure method.  I used to use send.firefox.com, but that's been discontinued.)

**Comment by kmsquire** [comment]
Minor status update: I originally thought that I had a runtime linking error, but it looks like we're calling `init_julia()` in `libcg.dll` fine.  It's the [call to `get_sysimg_path`](https://github.com/JuliaLang/PackageCompiler.jl/pull/490/files#diff-42a1f029580ad8620fc10c29f77540753a51dc21e020eebe19f087bfea61f541R54) (or something below that), which seems to be failing.  `JULIAC_PROGRAM_LIBNAME` should be set to `cg.dll` (at least, that's what appears in the PackageCompiler logs when compiling this package).

I'll debug more later this week.

**Comment by simonbyrne** [comment]
> I also need to set up a secret (NGROK_AUTH_TOKEN). Can I send that to you (maybe via Slack?) and ask you to set that for me under Settings -> Secrets -> Repository Secrets?
Sure, that's fine.

**Comment by simonbyrne** [comment]
@kmsquire is this working now?
https://github.com/simonbyrne/libcg/pull/18/checks?check_run_id=1964123892#step:7:5

**Comment by kmsquire** [comment]
Yes!  It turned out to be an issue with `dirname`.  On Windows, it modifies its input, which was the full path to the sysimage library, and which I was trying to reuse afterward.

I've already squashed it down there, but [here](https://github.com/JuliaLang/PackageCompiler.jl/pull/490/commits/56fbf69f1cb22104327fa6b98829a6ec8e39fba8)'s the PackageCompiler fix (on my branch).

**Comment by kmsquire** [comment]
The library size check is failing on Windows because I (mistakenly) thought that Windows libraries couldn't/shouldn't have `lib` as a prefix, but after looking into it, it seems like that's okay.  I'm going to remove that part of the `PackageCompiler` PR.

**Comment by simonbyrne** [comment]
Nice!

Should we add a check to see if it is relocateable?

**Comment by kmsquire** [comment]
Sure!  

FWIW, I've been able to use this successfully for work (on Mac and Linux)--I can build on one machine, package up the library and everything else required by Julia, install it under `/usr/local/lib`, and use it.  So relocatability shouldn't be a problem on those platforms.

For Windows, I think it's supposed to work similarly as long as the library location is in `PATH`, but I haven't been able to test that.

I'm going to spend some time rebasing and cleaning up the PackageCompiler PR first, then I can add a relocatability test here.  I might not get to that today.

**Comment by simonbyrne** [comment]
> For Windows, I think it's supposed to work similarly as long as the library location is in `PATH`, but I haven't been able to test that.

On Windows I believe that the binary directory is also in the search path: this is why many windows apps put the library in the `bin` directory, so then you don't need to modify `PATH`

**Comment by BambOoxX** [comment]
@kmsquire I may be overstepping here, but do you think `create_library` will allow to add the dll as a reference in a Visual Studio project ? I know you could not do full tests on windows, but that would be a very interesting development for me. 
The current `create_sysimage` exports functions decorated with the `Base.@ccallable` macro, these are visible using `dumpbin <nameofdll> /exports` on windows. However, if I want to link to this dll as a reference in Visual Studio, this fails.

**Comment by el-oso** [comment]
Don't you need to create an import libraries `.lib` file in Windows in order to be able to link the `.dll` ? 
You need to created this small static library in order to tell Windows the `.dll` size and stuff, perhaps that's why it is failing on Windows.

File: CG/LICENSE [hunk 3]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2021 Kevin Squire <kevin@secondspectrum.com> and contributors
```
Comment (simonbyrne) [inline CG/LICENSE:3]:
```suggestion
Copyright (c) 2021 Kevin Squire <kevin@secondspectrum.com>, Simon Byrne <simonbyrne@gmail.com> and contributors
```

Comment (simonbyrne) [reply CG/LICENSE:3]: @kmsquire Actually, there  is a license in the root directory: why 2 separate ones?

Comment (kmsquire) [reply CG/LICENSE]: This one was created by `PkgTemplates.jl` (see https://github.com/invenia/PkgTemplates.jl/pull/299).  I've removed this one int he latest push.

**Comment by kmsquire** [comment]
@BambOoxX, @el-oso, thank you for your comments, and sorry not to respond to them earlier.  I'll respond to each separately.

**Comment by kmsquire** [comment]
Okay, I just realized that @el-oso was responding to @BambOoxX--thank you!

FWIW, we actually did get Windows working, at least in C.  We recently added a rust example, but that only works on linux and OSX currently.

**Pull request merged**